feat(tests): Adds e2e tests for CSI

thomastaylor312 opened the pull request:
This PR adds e2e tests for the CSI code as part of our polishing for 1.0. Right now these tests only run on linux as some of the dependent components don't yet run on Mac. Also, the mock driver would need to use all of our ugly Windows socket code so we don't test it there either. Basically this is all a bit ugly, but it does exercise basic testing of the CSI components of the system. I also updated `podsmiter` to account for the new resources we are creating.

Closes #579

thomastaylor312 commented on Cargo.toml:
```diff
@@ -45,7 +45,7 @@ rustls-tls = [
 [dependencies]
 anyhow = "1.0"
-tokio = { version = "1.0", features = ["macros", "rt-multi-thread", "time"] }
+tokio = { version = "1.0", features = ["macros", "rt-multi-thread", "time", "process"] }
```
The new features and dependencies added here are for things in the unit tests and associated binaries (like podsmiter)

kflansburg replied:
Is this something that can be set in `dev-dependencies`?

thomastaylor312 replied:
Nope. There isn't a way to turn on the feature conditionally. I don't think you can have a duplicate dep in `dev-dependencies` can you?

kflansburg replied:
I'm pretty sure `kube` does this for `k8s-openapi` https://github.com/clux/kube-rs/blob/a6cc8995183a9ff9b0d7deb2ebad11dfa4ca31bb/kube/Cargo.toml#L86

thomastaylor312 replied:
Fixed!

thomastaylor312 commented on src/krustlet-wasi.rs:
```diff
@@ -15,6 +15,7 @@ async fn main() -> anyhow::Result<()> {
     // Initialize the logger
     tracing_subscriber::fmt()
         .with_writer(std::io::stderr)
+        .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
```
Yeah, we weren't using `RUST_LOG` properly. Whoops

kflansburg commented on tests/csi/README.md:
```diff
@@ -0,0 +1,24 @@
+# CSI Binaries directory
```
I believe that these should either leverage [git LFS](https://git-lfs.github.com/), or be hosted somewhere else. I would prefer not to download 70 MB when cloning the repository. (Maybe it already is and I just cant tell from the PR).

thomastaylor312 replied:
I didn't do this as it looked like it was gonna be a hassle. Let me give it a try

thomastaylor312 replied:
Wait, wouldn't this mean we'd have to set up LFS on every build run and anyone who wanted to run the tests locally would have to have it as well? I feel like I'd prefer storing the binary somewhere and downloading it on demand

kflansburg replied:
Hmmm, yes but this is pretty standard best practice, and easy to do. It’s pretty rude to have large binaries stored in git itself.

thomastaylor312 replied:
Yes I agree. But I think that downloading is preferable to having git LFS for 2 files only used during tests. If these were key components, maybe. But I hesitate to add Yet Another Step™ to getting someone going with krustlet

kflansburg replied:
Oh, I see, downloading is fine. I thought you were proposing keeping it in git.

kflansburg replied:
FWIW, I'm pretty sure if its stored in LFS, you can still clone the repo, you would only need to set up LFS to obtain the binaries.

thomastaylor312 replied:
Basically I am debating between having an on-demand GH action that builds these binaries and pushes them into the blobstore or just manually pushing them into the blob store. I'll probably revisit on Monday once I get some more random opinions from the internet. But I don't think git LFS is the right choice here. Removing them from git definitely is

bacongobbler replied:
It looks like [csi-release-tools](https://github.com/kubernetes-csi/csi-release-tools) only provides Docker images.

Have you considered pulling the binary out of the container image? That would alleviate the need to compile from source.

```
><> id=$(docker create k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.2.0)
><> docker cp $id:/csi-node-driver-registrar - | tar xf - > ./csi-node-driver-registrar
><> chmod 755 csi-node-driver-registrar
><> ./csi-node-driver-registrar --help
Usage of ./csi-node-driver-registrar:
  -add_dir_header
        If true, adds the file directory to the header of the log messages
```

thomastaylor312 replied:
That could work but adds a hard dependency on docker. I think pretty much everyone develops using `kind`, which means docker would be installed, but we haven't required that as of yet. I like keeping the set of tools small, but if this is easiest, we can do it this way

bacongobbler replied:
If you're hard set on either github actions vs git-lfs, I would prefer setting up a GitHub Action that builds the binaries from source, then pushes them up as a Github Release. That way we can keep CSI artifacts and Krustlet artifacts separate. Then in Krustlet we'd download those artifacts during a CI run.

You could set up a separate repository for that... csi-release-action would seem like a good descriptive name.

thomastaylor312 replied:
Ok, think I am going to keep it simple and have an on demand job that uploads them to the krustlet blobstore, purely for testing purposes. If people find having the compiled binaries around useful, we can break out a separate project

thomastaylor312 replied:
https://github.com/deislabs/krustlet/pull/614

thomastaylor312 replied:
Ok this is addressed

kflansburg commented on tests/test_resource_manager.rs:
```diff
@@ -180,16 +187,79 @@ impl TestResourceManager {
+            }
         }
 
+        //         apiVersion: v1
```
Should this be cleaned up?

thomastaylor312 replied:
Yeah, I'll take care of that shortly

kflansburg reviewed (changes requested):
The code itself looks reasonable, this looks like it was a huge effort!

I have a few administrative requests.

thomastaylor312 commented:
Ok @kflansburg and @bacongobbler. I added the new way of fetching the test binaries and squashed all the commits to remove the binaries from the branch

kflansburg reviewed (approved):
LGTM
